manifest/fusing_manifest_provider.go

```diff
@@ -0,0 +1,118 @@
+package manifest
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/filecoin-project/go-f3/ec"
+	"github.com/filecoin-project/go-f3/internal/clock"
+	"go.uber.org/multierr"
+	"golang.org/x/sync/errgroup"
+)
+
+// HeadGetter is the minimal subset of github.com/filecoin-project/go-f3/ec.Backend required by the
+// MixedManifestProvider.
+type HeadGetter interface {
+	GetHead(context.Context) (ec.TipSet, error)
+}
+
+var _ ManifestProvider = (*FusingManifestProvider)(nil)
+
+// FusingManifestProvider is a ManifestProvider that starts by providing dynamic manifest updates
+// then switches to a static manifest when we get within finality of said manifest's bootstrap
+// epoch.
+type FusingManifestProvider struct {
+	ec      HeadGetter
+	dynamic ManifestProvider
+	static  *Manifest
+
+	manifestCh chan *Manifest
+
+	errgrp     *errgroup.Group
+	cancel     context.CancelFunc
+	runningCtx context.Context
+	clock      clock.Clock
+}
+
+func NewFusingManifestProvider(ctx context.Context, ec HeadGetter, dynamic ManifestProvider, static *Manifest) (*FusingManifestProvider, error) {
+	if err := static.Validate(); err != nil {
+		return nil, err
+	}
+
+	clk := clock.GetClock(ctx)
+	ctx, cancel := context.WithCancel(context.WithoutCancel(ctx))
+	errgrp, ctx := errgroup.WithContext(ctx)
+
+	return &FusingManifestProvider{
+		ec:         ec,
+		dynamic:    dynamic,
+		static:     static,
+		errgrp:     errgrp,
+		cancel:     cancel,
+		runningCtx: ctx,
+		clock:      clk,
+		manifestCh: make(chan *Manifest, 1),
+	}, nil
+}
+
+func (m *FusingManifestProvider) ManifestUpdates() <-chan *Manifest {
+	return m.manifestCh
+}
+
+func (m *FusingManifestProvider) Start(ctx context.Context) error {
```

Comment:
I would add a bunch of logs here. I think we want to know what's happening in this function for debugging purposes, specifically around switch over epoch, when it actually happens and so on.

Reply:
I added a couple of logs.